rust/syncserver/src/lib.rs

```diff
@@ -0,0 +1,85 @@
+use std::net::{SocketAddr, TcpListener, TcpStream};
+use std::io::{BufReader, Read, Write, Error, BufWriter};
+use std::thread;
+
+const MAX_MESSAGE_SIZE: usize = 128;
+const NUM_CLIENTS: i32 = 10;
+
+/// Trait defining behavior for the server/client
+pub trait Runnable {
+
+    /// Start the server/client. The implementation of this function is expected to loop infinitely
+    fn start(&self) -> Result<(), Error>;
+
+    /// Perform one send/receive cycle with the other end of the socket.
+    /// Return the TcpStream if it's still open, create an error otherwise.
+    fn handle_connection(thread_id: i32, stream: TcpStream) -> Result<(), Error>;
+}
+
+// TcpServer: Start the listener and delegate incoming connections to other threads.
+
+pub struct TcpServer {
+    pub socket: SocketAddr
+}
+
+// Runnable handler for the TCP server
+impl Runnable for TcpServer {
+    fn start(&self) -> Result<(), Error> {
+        let listener = TcpListener::bind(self.socket)?;
+        let mut thread_id = 0;
+        for stream in listener.incoming() {
+            let stream = stream.unwrap();
+            thread::spawn(move || TcpServer::handle_connection(thread_id, stream));
```

Comment:
from the [docs](https://doc.rust-lang.org/std/keyword.move.html):
```
move converts any variables captured by reference or mutable reference to variables captured by value.
```
So actually, I believe I need the `move` here for thread_id, not for stream. Once stream is passed to handle_connection, the scope for handle_connection takes ownership of the stream. Here, thread_id is the problem since this scope maintains a mutable reference to it. See the [Copy](https://doc.rust-lang.org/std/marker/trait.Copy.html) trait for more.

Here's an example of this:

```

fn main() {
    let mut v = vec![1,2,3];
    let mut reg_closure = |x| v.push(x);
    reg_closure(4);
    println!("{:?}", v);
}

```
the above code will still compile because the closure has a mutable reference to `v`, so when it's invoked in this scope, it will alter the underlying memory and the return ownership of `v` back to the scope of `main()` after the `push` operation is completed.

In contrast:
```
fn main() {
    let x = 5;
    test_i32(x);
    println!("{x}");

    let mut v = vec![1,2,3];
    let mut move_closure = move |x| v.push(x);
    move_closure(4);
    println!("{:?}", v);
}
```
will fail to compile because we try to print v after we've moved ownership to another scope. But if we took the print line out, this would compile fine. since the closure's scope now owns `v`. In the case of `thread_id` in my code, i32 values will be copied by value upon a move. So the value is fully copied over to a new address on the stack and the mutable reference to thread_id is maintained in the handle_connection() scope.